File: Python/GrailSort.py
# additional functions
def compareVal(a, b):
    return (a > b) - (a < b) 

def arrayCopy(fromArray, fromIndex, toArray, toIndex, length):
    toArray[toIndex:toIndex + length] = fromArray[fromIndex:fromIndex + length]


 #
 # MIT License
 # 
 # Copyright (c) 2013 Andrey Astrelin
 # Copyright (c) 2020 The Holy Grail Sort Project
 # 
 # Permission is hereby granted, free of charge, to any person obtaining a copy
 # of this software and associated documentation files (the "Software"), to deal
 # in the Software without restriction, including without limitation the rights
 # to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 # copies of the Software, and to permit persons to whom the Software is
 # furnished to do so, subject to the following conditions:
 # 
 # The above copyright notice and this permission notice shall be included in all
 # copies or substantial portions of the Software.
 # 
 # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 # IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 # FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 # AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 #
 #
 # The Holy Grail Sort Project
 # Project Manager:      Summer Dragonfly
 # Project Contributors: 666666t
 #                       Anonymous0726
 #                       aphitorite
 #                       dani_dlg
 #                       EilrahcF
 #                       Enver
 #                       lovebuny
 #                       MP
 #                       phoenixbound
 #                       thatsOven
 #                       
 # Special thanks to "The Studio" Discord community!
 #
class Subarray:
    LEFT = 0
    RIGHT = 0

# REWRITTEN GRAILSORT FOR PYTHON - A heavily refactored C/C++-to-Python version of
#                                  Andrey Astrelin's GrailSort.h, aiming to be as
#                                  readable and intuitive as possible.
#
# ** Written and maintained by The Holy Grail Sort Project
#
# Primary author: thatsOven
#
# Current status: Not finished yet! 

class GrailSort:

    externalBuffer = [] #TODO Check these definitions
    externalBufferLength = 0
    currentBlockLen = 0

    def __init__(self, grailKeys):
        self.grailKeys = grailKeys
        self.staticExternalBufferLen = 512

    def grailSwap(self, array, a, b):
        array[a], array[b] = array[b], array[a]
    
    def grailBlockSwap(self, array, a, b, swapsLeft):
        swaps = 0
        for _ in range(swapsLeft, 0, -1):
            self.grailSwap(array, a + swaps, b + swaps)
            swaps += 1

    def grailRotate(self, array, start, leftLen, rightLen):
        while leftLen > 0 and rightLen > 0:
            if leftLen <= rightLen:
                self.grailBlockSwap(array, start, start + leftLen, leftLen)
                start += leftLen
                rightLen -= leftLen
            else:
                self.grailBlockSwap(array, start + leftLen - rightLen, start + leftLen, rightLen)
                leftLen -= rightLen
    
    def grailBinarySearchLeft(self, array, start, length, target):
        left = 0
        right = length
        while left < right:
            middle = left + ((right - left) // 2)
            if array[start + middle] < target:
                left = middle + 1
            else:
                right = middle
        return left

    def grailBinarySearchRight(self, array, start, length, target):
        left = 0
        right = length
        while left < right:
            middle = left + ((right - left) // 2)
            if array[start + middle] > target:
                right = middle
            else:
                left = middle + 1
        return right

    def grailCollectKeys(self, array, start, length, idealKeys):
        keysFound = 1
        firstKey = 0
        currentKey = 1

        while currentKey < length and keysFound < idealKeys:

            insertPos = self.grailBinarySearchLeft(array, start + firstKey, keysFound, array[start + currentKey].key)

            if insertPos == keysFound or not array[start + currentKey].key == array[start + firstKey + insertPos].key:

                self.grailRotate(array, start + firstKey, keysFound, currentKey - (firstKey + keysFound))

                firstKey = currentKey - keysFound

                self.grailRotate(array, start + firstKey + insertPos, keysFound - insertPos, 1)

                keysFound += 1

            currentKey += 1

        self.grailRotate(array, start, firstKey, keysFound)
        return keysFound

    def grailPairwiseSwaps(self, array, start, length):
        for index in range(1, length, 2):
            left = start + index - 1
            right = start + index

            if array[left].key > array[right].key:
                self.grailSwap(array,  left - 2, right)
                self.grailSwap(array, right - 2,  left)
            else:
                self.grailSwap(array,  left - 2,  left)
                self.grailSwap(array, right - 2, right)
        
        left = start + index - 1
        if left < start + length:
            self.grailSwap(array, left - 2, left)

    def grailPairwiseWrites(self, array, start, length):
        for index in range(1, length, 2):
            left = start + index - 1
            right = start + index

            if array[left] > array[right]:
                array[left - 2], array[right - 2] = array[right], array[left]
            else:
                array[left - 2], array[right - 2] = array[left], array[right]

            left = start + index - 1
            if left < start + length:
                array[left - 2] = array[left]

    def grailBlockSelectSort(self, array, keys, start, medianKey, blockCount, blockLen):
        for block in range(1, blockCount):
            left = block - 1
            right = left

            for index in range(block, blockCount):
                compare = compareVal(   array[start + (right * blockLen)], 
                                        array[start + (index * blockLen)]   )
                
                if compare > 0 or (compare == 0 and compareVal( array[keys + right], 
                                                                array[keys + index]) > 0 ):
                    right = index

                if right != left:

                    self.grailBlockSwap(array, start + (left * blockLen), start + (right * blockLen), blockLen)

                    self.grailSwap(array, keys + left, keys + right)

                    if medianKey == left: medianKey = right
                    elif medianKey == right: medianKey = left

        return medianKey

    def grailMergeForwards(self, array, start, leftLen, rightLen, bufferOffset):
        left   = start
        middle = start + leftLen
        right  = middle
        end    = middle + rightLen
        buffer = start - bufferOffset

        while right < end:
            if left == middle or array[left] > array[right]:

                self.grailSwap(array, buffer, right)
                right += 1
            else:
                self.grailSwap(array, buffer, left)
                left += 1
            buffer += 1

            if buffer != left:
                self.grailBlockSwap(array, buffer, left, middle-left)
                                
    def grailMergeBackwards(self, array, start, leftLen, rightLen, bufferOffset):
        left   = start + leftLen - 1
        middle = left
        right  = middle + rightLen
        end    = start
        buffer = right + bufferOffset

        while left >= end:
            if right == middle or array[left] > array[right]:

                self.grailSwap(array, buffer, left)
                left -= 1
            else:
                self.grailSwap(array, buffer, right)
                right -= 1
            buffer -= 1

        if right != buffer:
            while right > middle:
                self.grailSwap(array, buffer, right)
                buffer -= 1
                right -= 1
        
    def grailOutOfPlaceMerge(self, array, start, leftLen, rightLen, bufferOffset):
        left   = start
        middle = start + leftLen
        right  = middle
        end    = middle + rightLen
        buffer = start - bufferOffset
        
        while right < end:
            if left == middle or array[left] > array[right]:

                array[buffer] = array[right]
                right += 1
            else:
                array[buffer] = array[left]
                left += 1
            buffer += 1

        if buffer != left:
            while left < middle:
                array[buffer] = array[left]
                buffer += 1
                left += 1
    
    def grailInPlaceBufferReset(self, array, start, resetLen, bufferLen):
        index = start + resetLen - 1
        while index >= 0:
            self.grailSwap(array, start + index, start + index - bufferLen)
            index -= 1

    def grailOutOfPlaceBufferReset(self, array, start, resetLen, bufferLen):
        index = start + resetLen
        while index >= 0:
            array[index] = array[index - bufferLen]
            index -= 1

    def grailInPlaceBufferRewind(self, array, start, buffer, leftOvers):
        while start < buffer:
            buffer -= 1
            leftOvers -= 1
            self.grailSwap(array, buffer, leftOvers)

    def grailOutOfPlaceBufferRewind(self, array, start, buffer, leftOvers):
        while start < buffer:
            buffer -= 1
            leftOvers -= 1
            array[buffer] = array[leftOvers]

    def grailBuildBlocks(self, array, start, length, bufferLen):
        if self.externalBuffer is not None:
            if bufferLen < self.externalBufferLength:
                externLen = bufferLen
            else:
                externLen = 1
                while (externLen*2) <= self.externalBufferLength:
                    externLen *= 2
            self.grailBuildOutOfPlace(array, start, length, bufferLen, externLen)
        else:
            self.grailPairwiseSwaps(array, start, length)
            self.grailBuildInPlace(array, start - 2, length, 2, bufferLen)
        
    def grailBuildOutOfPlace(self, array, start, length, bufferLen, externLen):
        arrayCopy(array, start - externLen, self.externalBuffer, 0, externLen)

        self.grailPairwiseWrites(array, start, length)
        start -= 2

        mergeLen = 2
        while mergeLen < externLen:
            mergeEnd = start + length - (2 * mergeLen)
            bufferOffset = mergeLen

            mergeIndex = start
            while mergeIndex <= mergeEnd:
                self.grailOutOfPlaceMerge(array, mergeIndex, mergeLen, mergeLen, bufferOffset)
                mergeIndex += (2 * mergeLen)

            leftOver = length - (mergeIndex - start)

            if leftOver > mergeLen:
                self.grailOutOfPlaceMerge(array, mergeIndex, mergeLen, leftOver - mergeLen, bufferOffset)
            else:
                #TODO: Is this correct?? (from Java code)
                for offset in range(0, leftOver):
                    array[mergeIndex + offset - mergeLen] = array[mergeIndex + offset]
            
            start -= mergeLen
            mergeLen *= 2

        arrayCopy(self.externalBuffer, 0, array, start+length, externLen)
        self.grailBuildInPlace(array, start, length, mergeLen, bufferLen)

    def grailBuildInPlace(self, array, start, length, currentMerge, bufferLen):
        mergeLen = currentMerge
        while mergeLen < bufferLen:
            mergeEnd = start + length - (2 * mergeLen)
            bufferOffset = mergeLen

            mergeIndex = start
            while mergeIndex <= mergeEnd:
                self.grailMergeForwards(array, mergeIndex, mergeLen, mergeLen, bufferOffset)
                mergeIndex += (2 * mergeLen)

            leftOver = length - (mergeIndex - start)

            if leftOver > mergeLen:
                self.grailMergeForwards(array, mergeIndex, mergeLen, leftOver - mergeLen, bufferOffset)
            else:
                self.grailRotate(array, mergeIndex - mergeLen, mergeLen, leftOver)

            start -= mergeLen
            mergeLen *= 2
        
        finalBlock = length % (2 * bufferLen)
        finalOffset = start + length - finalBlock

        if finalBlock <= bufferLen:
            self.grailRotate(array, finalOffset, finalBlock, bufferLen)
        else:
            self.grailMergeBackwards(array, finalOffset, bufferLen, finalBlock - bufferLen, bufferLen)

        mergeIndex = finalOffset - (2 * bufferLen)
        while mergeIndex >= start:
            self.grailMergeBackwards(array, mergeIndex, bufferLen, bufferLen, bufferLen)
            mergeIndex -= (2 * bufferLen)

    def grailCountLeftBlocks(self, array, offset, blockCount, blockLen):
        leftBlocks = 0

        firstRightBlock = offset + (blockCount * blockLen)
        prevLeftBlock   = firstRightBlock - blockLen

        while (leftBlocks < blockCount) and (array[firstRightBlock] < array[prevLeftBlock]):
            leftBlocks += 1
            prevLeftBlock -= blockLen

        return leftBlocks

    def grailGetSubarray(self, array, currentKey, medianKey):
        if array[currentKey] < array[medianKey] : return Subarray.LEFT
        else                                    : return Subarray.RIGHT

    def grailSmartMergeOutOfPlace(self, array, start, leftLen, leftOrigin: Subarray, rightLen, bufferOffset):
        left   = start
        middle = start + leftLen
        right  = middle
        end    = middle + rightLen
        buffer = start - bufferOffset

        if leftOrigin == Subarray.LEFT:
            while (left < middle) and (right < end):
                if array[left] <= array[right]:
                    array[buffer] = array[left]
                    left += 1
                else:
                    array[buffer] = array[right]
                    right += 1
                buffer += 1

        else:
            while (left < middle) and (right < end):
                if array[left] < array[right]:
                    array[buffer] = array[left]
                    left += 1
                else:
                    array[buffer] = array[right]
                    right += 1
                buffer += 1

        if left < middle:
            self.currentBlockLen = middle - left
            self.grailOutOfPlaceBufferRewind(array, left, middle, end)
        else:
            self.currentBlockLen = end - right
            if leftOrigin == Subarray.LEFT:
                self.currentBlockOrigin = Subarray.RIGHT
            else:
                self.currentBlockOrigin = Subarray.LEFT

    def grailSmartMerge(self, array, start, leftLen, leftOrigin, rightLen, bufferOffset):
        left   = start
        middle = start + leftLen
        right  = middle
        end    = middle + rightLen
        buffer = start - bufferOffset

        if leftOrigin == Subarray.LEFT:
            while (left < middle) and (right < end):
                if array[left] <= array[right]:
                    self.grailSwap(array, buffer, left)
                    left += 1
                else:
                    self.grailSwap(array, buffer, right)
                    right += 1
                buffer += 1
        
        else:
            while (left < middle) and (right < end):
                if array[left] < array[right]:
                    self.grailSwap(array, buffer, left)
                    left += 1
                else:
                    self.grailSwap(array, buffer, right)
                    right += 1
                buffer += 1

        if left < middle:
            self.currentBlockLen = middle - left
            self.grailInPlaceBufferRewind(array, left, middle, end)
        else:
            self.currentBlockLen = end - right
            if leftOrigin < Subarray.LEFT:
                self.currentBlockOrigin = Subarray.RIGHT
            else:
                self.currentBlockOrigin = Subarray.LEFT

    def grailSmartLazyMerge(self, array, start, leftLen, leftOrigin, rightLen):
        if leftOrigin == Subarray.LEFT:
            if array[start + leftLen - 1] > array[start + leftLen]:
                while leftLen != 0:
                    insertPos = self.grailBinarySearchLeft(array, start + leftLen, rightLen, array[start])

                    if insertPos != 0:
                        self.grailRotate(array, start, leftLen, insertPos)
                        start    += insertPos
                        rightLen -= insertPos

                    if rightLen == 0:
                        self.currentBlockLen = leftLen
                        return
                    else:
                        condition = True
                        while condition:
                            start += 1
                            leftLen -= 1
                            condition = (leftLen != 0) and array[start] <= array[start + leftLen]
        else:
            if array[start + leftLen - 1] >= array[start + leftLen]:
                while leftLen != 0:
                    insertPos = self.grailBinarySearchRight(array, start + leftLen, rightLen, array[start])

                    if insertPos != 0:
                        self.grailRotate(array, start, leftLen, insertPos)
                        start    += insertPos
                        rightLen -= insertPos

                    if rightLen == 0:
                        self.currentBlockLen = leftLen
                        return
                    else:
                        condition = True
                        while condition:
                            start += 1
                            leftLen -= 1
                            condition = (leftLen !=  0) and (array[start] < array[start + leftLen])
        
        self.currentBlockLen = rightLen
        if leftOrigin == Subarray.LEFT:
            self.currentBlockOrigin = Subarray.RIGHT
        else:
            self.currentBlockOrigin = Subarray.LEFT

    
        

File: C/Gaming32's Rewritten Grailsort for C/grail_sort.c
static void FUNC(grailBlockSwap)(VAR* a, VAR* b, size_t blockLen) {
    for (size_t i = 0; i < blockLen; i++) {
        VAR tmp = *a;
        *a++ = *b;
        *b++ = tmp;
    }
}

static void FUNC(grailRotate)(VAR* start, size_t leftLen, size_t rightLen) {
    while (leftLen > 0 && rightLen > 0) {
        if (leftLen <= rightLen) {
            FUNC(grailBlockSwap)(start, start + leftLen, leftLen);
            start    += leftLen;
            rightLen -= leftLen;
        } else {
            FUNC(grailBlockSwap)(start + leftLen - rightLen, start + leftLen, rightLen);
            leftLen -= rightLen;
        }
    }
}

static void FUNC(grailInsertSort)(VAR* start, VAR* end, GRAILCMP cmp) {
    for (VAR* item = start + 1; item < end; item++) {
        VAR* left = item - 1;
        VAR* right = item;

        while (left >= start && cmp(left, right) > 0) {
            VAR tmp = *left;
            *left-- = *right;
            *right-- = tmp;
        }
    }
}

static size_t FUNC(grailBinarySearchLeft)(VAR* start, size_t length, VAR* target, GRAILCMP cmp) {
    size_t  left = 0;
    size_t right = length;

    while (left < right) {
        size_t middle = left + ((right - left) / 2);

        if (cmp(start + middle, target) < 0) {
            left = middle + 1;
        } else {
            right = middle;
        }
    }
    return left;
}

static size_t FUNC(grailBinarySearchRight)(VAR* start, size_t length, VAR* target, GRAILCMP cmp) {
    size_t  left = 0;
    size_t right = length;

    while (left < right) {
        size_t middle = left + ((right - left) / 2);

        if (cmp(start + middle, target) > 0) {
            right = middle;
        } else {
            left = middle + 1;
        }
    }

    return right;
}

static size_t FUNC(grailCollectKeys)(VAR* start, VAR* end, size_t idealKeys, GRAILCMP cmp) {
    size_t keysFound = 1;
    VAR*    firstKey = start;
    VAR*     currKey = start + 1;

    while (currKey < end && keysFound < idealKeys) {
        size_t insertPos = FUNC(grailBinarySearchLeft)(firstKey, keysFound, currKey, cmp);

        if (insertPos == keysFound || cmp(currKey, firstKey + insertPos) != 0) {
            FUNC(grailRotate)(firstKey, keysFound, currKey - (firstKey + keysFound));

            firstKey = currKey - keysFound;

            VAR tmp = firstKey[keysFound];
            memmove(firstKey + insertPos + 1, firstKey + insertPos, sizeof(VAR) * (keysFound - insertPos));
            firstKey[insertPos] = tmp;

            keysFound++;
        }

        currKey++;
    }

    FUNC(grailRotate)(start, firstKey - start, keysFound);
    return keysFound;
}

static void FUNC(grailPairwiseSwaps)(VAR* start, size_t length, GRAILCMP cmp) {
    VAR tmp;
    size_t index;
    for (index = 1; index < length; index += 2) {
        VAR*  left = start + index - 1;
        VAR* right = start + index;

        if (cmp(left, right) > 0) {
            tmp = left[-2];
            left[-2] = *right;
            *right = tmp;

            tmp = right[-2];
            right[-2] = *left;
            *left = tmp;
        } else {
            tmp = left[-2];
            left[-2] = *left;
            *left = tmp;

            tmp = right[-2];
            right[-2] = *right;
            *right = tmp;
        }
    }

    VAR* left = start + index - 1;
    if (left < start + length) {
        tmp = left[-2];
        left[-2] = *left;
        *left = tmp;
    }
}

static void FUNC(grailPairwiseWrites)(VAR* start, size_t length, GRAILCMP cmp) {
    size_t index;
    for (index = 1; index < length; index += 2) {
        VAR*  left = start + index - 1;
        VAR* right = start + index;

        if (cmp(left, right) > 0) {
            left[ -2] = *right;
            right[-2] = *left;
        } else {
            left[ -2] = *left;
            right[-2] = *right;
        }
    }

    VAR* left = start + index - 1;
    if (left < start + length) {
        left[-2] = *left;
    }
}

static void FUNC(grailMergeForwards)(VAR* start, size_t leftLen, size_t rightLen, size_t bufferOffset, GRAILCMP cmp) {
    VAR* buffer = start  - bufferOffset;
    VAR*   left = start;
    VAR* middle = start  + leftLen;
    VAR*  right = middle;
    VAR*    end = middle + rightLen;

    VAR tmp;
    while (right < end) {
        if (left == middle || cmp(left, right) > 0) {
            tmp = *buffer;
            *buffer++ = *right;
            *right++ = tmp;
        } else {
            tmp = *buffer;
            *buffer++ = *left;
            *left++ = tmp;
        }
    }

    if (buffer != left) {
        FUNC(grailBlockSwap)(buffer, left, middle - left);
    }
}

static void FUNC(grailMergeBackwards)(VAR* start, size_t leftLen, size_t rightLen, size_t bufferOffset, GRAILCMP cmp) {
    VAR*    end = start - 1;
    VAR*   left = end   + leftLen;
    VAR* middle = left;
    VAR*  right = middle + rightLen;
    VAR* buffer = right + bufferOffset;

    VAR tmp;
    while (left > end) {
        if (right == middle || cmp(left, right) > 0) {
            tmp = *buffer;
            *buffer-- = *left;
            *left-- = tmp;
        } else {
            tmp = *buffer;
            *buffer-- = *right;
            *right-- = tmp;
        }
    }

    if (right != buffer) {
        while (right > middle) {
            tmp = *buffer;
            *buffer-- = *right;
            *right-- = tmp;
        }
    }
}

static void FUNC(grailMergeOutOfPlace)(VAR* start, size_t leftLen, size_t rightLen, size_t bufferOffset, GRAILCMP cmp) {
    VAR* buffer = start  - bufferOffset;
    VAR*   left = start;
    VAR* middle = start  + leftLen;
    VAR* right = middle;
    VAR*   end = middle + rightLen;

    while (right < end) {
        if (left == middle || cmp(left, right) > 0) {
            *buffer++ = *right++;
        } else {
            *buffer++ = *left++;
        }
    }

    if (buffer != left) {
        while (left < middle) {
            *buffer++ = *left++;
        }
    }
}

static void FUNC(grailBuildInPlace)(VAR* start, size_t length, size_t currentLen, size_t bufferLen, GRAILCMP cmp) {
    for (size_t mergeLen = currentLen; mergeLen < bufferLen; mergeLen *= 2) {
        size_t fullMerge = 2 * mergeLen;

        VAR* mergeIndex;
        VAR* mergeEnd = start + length - fullMerge;
        size_t bufferOffset = mergeLen;

        for (mergeIndex = start; mergeIndex <= mergeEnd; mergeIndex += fullMerge) {
            FUNC(grailMergeForwards)(mergeIndex, mergeLen, mergeLen, bufferOffset, cmp);
        }

        size_t leftOver = length - (mergeIndex - start);

        if (leftOver > mergeLen) {
            FUNC(grailMergeForwards)(mergeIndex, mergeLen, leftOver - mergeLen, bufferOffset, cmp);
        } else {
            FUNC(grailRotate)(mergeIndex - mergeLen, mergeLen, leftOver);
        }

        start -= mergeLen;
    }

    size_t fullMerge  = 2 * bufferLen;
    size_t lastBlock  = length & (fullMerge - 1); // equivalent to modulo, but faster (only works if fullMerge is a power of 2, which it is)
    VAR*   lastOffset = start + length - lastBlock;

    if (lastBlock <= bufferLen) {
        FUNC(grailRotate)(lastOffset, lastBlock, bufferLen);
    } else {
        FUNC(grailMergeBackwards)(lastOffset, bufferLen, lastBlock - bufferLen, bufferLen, cmp);
    }

    for (VAR* mergeIndex = lastOffset - fullMerge; mergeIndex >= start; mergeIndex -= fullMerge) {
        FUNC(grailMergeBackwards)(mergeIndex, bufferLen, bufferLen, bufferLen, cmp);
    }
}

static void FUNC(grailBuildOutOfPlace)(VAR* start, size_t length, size_t bufferLen, VAR* extBuffer, size_t extLen, GRAILCMP cmp) {
    memcpy(extBuffer, start - extLen, sizeof(VAR) * extLen);

    FUNC(grailPairwiseWrites)(start, length, cmp);
    start -= 2;

    size_t mergeLen;
    for (mergeLen = 2; mergeLen < extLen; mergeLen *= 2) {
        size_t fullMerge = 2 * mergeLen;

        VAR* mergeIndex;
        VAR* mergeEnd = start + length - fullMerge;
        size_t bufferOffset = mergeLen;

        for (mergeIndex = start; mergeIndex <= mergeEnd; mergeIndex += fullMerge) {
            FUNC(grailMergeOutOfPlace)(mergeIndex, mergeLen, mergeLen, bufferOffset, cmp);
        }

        size_t leftOver = length - (mergeIndex - start);

        if (leftOver > mergeLen) {
            FUNC(grailMergeOutOfPlace)(mergeIndex, mergeLen, leftOver - mergeLen, bufferOffset, cmp);
        }
        else {
            memmove(mergeIndex - mergeLen, mergeIndex, sizeof(VAR) * leftOver);
        }

        start -= mergeLen;
    }

    memcpy(start + length, extBuffer, sizeof(VAR) * extLen);
    FUNC(grailBuildInPlace)(start, length, mergeLen, bufferLen, cmp);
}

static void FUNC(grailBuildBlocks)(VAR* start, size_t length, size_t bufferLen, VAR* extBuffer, size_t extBufferLen, GRAILCMP cmp) {
    if (extBuffer != NULL) {
        size_t extLen;

        if (bufferLen < extBufferLen) {
            extLen = bufferLen;
        } else {
            extLen = 1;
            while ((extLen * 2) <= extBufferLen) {
                extLen *= 2;
            }
        }

        FUNC(grailBuildOutOfPlace)(start, length, bufferLen, extBuffer, extLen, cmp);
    } else {
        FUNC(grailPairwiseSwaps)(start, length, cmp);
        FUNC(grailBuildInPlace)(start - 2, length, 2, bufferLen, cmp);
    }
}

static void FUNC(grailLazyMerge)(VAR* start, size_t leftLen, size_t rightLen, GRAILCMP cmp) {
    if (leftLen < rightLen) {
        VAR* middle = start + leftLen;

        while (leftLen != 0) {
            size_t mergeLen = FUNC(grailBinarySearchLeft)(middle, rightLen, start, cmp);

            if (mergeLen != 0) {
                FUNC(grailRotate)(start, leftLen, mergeLen);

                start    += mergeLen;
                middle   += mergeLen;
                rightLen -= mergeLen;
            }

            if (rightLen == 0) {
                break;
            } else {
                do {
                    start++;
                    leftLen--;
                } while (leftLen != 0 && cmp(start, middle) <= 0);
            }
        }
    } else {
        VAR* end = start + leftLen + rightLen - 1;

        while (rightLen != 0) {
            size_t mergeLen = FUNC(grailBinarySearchRight)(start, leftLen, end, cmp);

            if (mergeLen != leftLen) {
                FUNC(grailRotate)(start + mergeLen, leftLen - mergeLen, rightLen);

                end    -= leftLen - mergeLen;
                leftLen = mergeLen;
            }

            if (leftLen == 0) {
                break;
            } else {
                VAR* middle = start + leftLen;
                do {
                    rightLen--;
                    end--;
                } while (rightLen != 0 && cmp(middle - 1, end) <= 0);
            }
        }
    }
}

static void FUNC(grailLazyStableSort)(VAR* start, size_t length, GRAILCMP cmp) {
    VAR* end = start + length;
    for (VAR *left = start, *right = start + 1; right < end; left += 2, right += 2) {
        if (cmp(left, right) > 0) {
            VAR tmp = *left;
            *left = *right;
            *right = tmp;
        }
    }
    for (size_t mergeLen = 2; mergeLen < length; mergeLen *= 2) {
        size_t fullMerge = 2 * mergeLen;

        size_t mergeIndex;
        size_t mergeEnd = length - fullMerge + 1;

        for (mergeIndex = 0; mergeIndex <= mergeEnd; mergeIndex += fullMerge) {
            FUNC(grailLazyMerge)(start + mergeIndex, mergeLen, mergeLen, cmp);
        }

        size_t leftOver = length - mergeIndex;
        if (leftOver > mergeLen) {
            FUNC(grailLazyMerge)(start + mergeIndex, mergeLen, leftOver - mergeLen, cmp);
        }
    }
}

static void FUNC(grailCommonSort)(VAR* start, size_t length, VAR* extBuffer, size_t extBufferLen, GRAILCMP cmp) {
    VAR* end = start + length;
    if (length < 16) {
        FUNC(grailInsertSort)(start, end, cmp);
        return;
    }

    size_t blockLen = 1;

    while ((blockLen * blockLen) < length) {
        blockLen *= 2;
    }

    size_t keyLen = ((length - 1) / blockLen) + 1;

    size_t idealKeys = keyLen + blockLen;

    size_t keysFound = FUNC(grailCollectKeys)(start, end, idealKeys, cmp);

    bool idealBuffer;
    if (keysFound < idealKeys) {
        if (keysFound < 4) {
            FUNC(grailLazyStableSort)(start, length, cmp);
            return;
        } else {
            keyLen = blockLen;
            blockLen = 0;
            idealBuffer = false;

            while (keyLen > keysFound) {
                keyLen /= 2;
            }
        }
    } else {
        idealBuffer = true;
    }

    size_t bufferEnd = blockLen + keyLen;
    size_t subarrayLen;
    if (idealBuffer) {
        subarrayLen = blockLen;
    } else {
        subarrayLen = keyLen;
    }

    FUNC(grailBuildBlocks)(start + bufferEnd, length - bufferEnd, subarrayLen, extBuffer, extBufferLen, cmp);
    return;

    while ((length - bufferEnd) > (2 * subarrayLen)) {
        subarrayLen *= 2;

        size_t currentBlockLen = blockLen;
        bool scrollingBuffer = idealBuffer;

        if (!idealBuffer) {
            size_t keyBuffer = keyLen / 2;

            if (keyBuffer >= ((2 * subarrayLen) / keyBuffer)) {
                currentBlockLen = keyBuffer;
                scrollingBuffer = true;
            }
            else {
                currentBlockLen = (2 * subarrayLen) / keyLen;
            }
        }
    }
}
